feat: rewrite the scanner in C, fix some infinite loop bugs

amaanq opened the pull request:
Somewhat a follow-up to https://github.com/Kerl13/tree-sitter-menhir/pull/5, I have now replaced many existing scanners in C++ to C - the logic for variable sized strings works well and is shared across multiple tree-sitter libraries (I'm thinking of maybe adding a common string/vector helper library to tree-sitter core for that..)

314eter commented:
I still have the same reservations as for Kerl13/tree-sitter-menhir#5.

- Can you point to any example/discussion where C++ scanners create problems? This rewrite adds complexity, which should be justified. This doesn't mean I don't want to accept a rewrite, I just want to be sure there is a good reason to do it. If someone comes along in the future with a "use C++ stdlib" PR, why should we reject it?
- You mention bugs in the title, but not in the description. If you fix bugs, please clearly point out the bug and the fix. Preferably with an example input that we can add to the tests.
- It's not just a rewrite, you made some other changes too. Some of them look accidental, and some of them don't. It's difficult to review if you don't separate the rewrite from the other changes.
- There are several failing tests.

314eter commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
 
-      while (lexer->lookahead == ' ' || lexer->lookahead == '\t') advance(lexer);
+String string_new() {
+    return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)};
```
Should we `assert` that `calloc()` doesn't return `NULL`? If you do it in `STRING_RESIZE()`, it's weird not to do it here.

amaanq replied:
hmm yeah, that's a good idea

314eter commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
 
-      while (lexer->lookahead != '\n' && lexer->lookahead != '\r') advance(lexer);
+unsigned serialize(Scanner *scanner, char *buffer) {
+    return 0;
```
Why is this return here?

amaanq replied:
Debugging...wow oops :sweat_smile:

314eter commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
-    } else if (valid_symbols[NULL_CHARACTER] && lexer->lookahead == '\0') {
-      return !lexer->eof(lexer);
+    if (size + 1 > TREE_SITTER_SERIALIZATION_BUFFER_SIZE) {
+        return 0;
```
Isn't it better to crash instead? Returning 0 will result in weird behaviour,

amaanq replied:
No, it's common in most scanners to just truncate extra data, crashing wouldn't be ideal - same principle applies in python, bash, etc

314eter replied:
tree-sitter-cpp is [crashing](https://github.com/tree-sitter/tree-sitter-cpp/blob/0e7b7a02b6074859b51c1973eb6a8275b3315b1d/src/scanner.c#L129-L131). It would be possible to only save the start of `quoted_string_id`, but that complicates the implementation, and in practice `quoted_string_id` will never be that long.

amaanq replied:
Hmm yeah, I guess crashing could be fine here too

amaanq commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
-  NULL_CHARACTER
-};
+#define STRING_RESIZE(vec, _cap)                                               \
+    void *tmp = realloc((vec).data, ((_cap) + 1) * sizeof((vec).data[0]));     \
```
Null terminator

314eter replied:
But we're not using null-terminated strings here. You know the length of a string becaused it's stored in `len`.

amaanq replied:
Yes, but C library functions want that - in the context of this scanner, sure, we don't. But say someone uses strlen or strcpy in the future for the grammar, this would cause issues then

314eter replied:
Using `strlen()` here would be wrong. If you need the length of `quoted_string_id`, you should use the value stored in `len`. You're adding a lot of complexity to support something that I'd consider to be a bug.

amaanq replied:
yeah, true, I guess strcmp/strcpy are more valid reasons but I'll remove that anyways

amaanq commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
+        lexer->get_column(lexer) == 0) {
+        advance(lexer);
+
+        while (iswspace(lexer->lookahead)) {
```
yes, my bad, I did that because theres other kinds of spaces a user might use, unless that's not valid in ocaml (but I'd assume it is)

314eter replied:
We're matching [line number directives](https://v2.ocaml.org/manual/lex.html#sss:lex-linedir) here. They are used in code generated by preprocessors, to refer to the original source code, and look like this:
```ocaml
# 123 "file.mll"
```
They can only occur at the beginning of a line (which is why we need an external scanner), and can't span multiple lines. So this is not valid:
```ocaml
#
123
"file.mll"
```

amaanq replied:
Gotcha - I meant special spaces defined in here:
https://jkorpela.fi/chars/spaces.html

Though I updated it to not count \r and \n

314eter replied:
Only [spaces and tabs](https://github.com/ocaml/ocaml/blob/182367e101d8dfcc87fd271fffe7b6b53211a1d7/parsing/lexer.mll#L594-L596).

amaanq replied:
Thanks! Fixed

amaanq commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
-  NULL_CHARACTER
-};
+#define STRING_RESIZE(vec, _cap)                                               \
+    void *tmp = realloc((vec).data, ((_cap) + 1) * sizeof((vec).data[0]));     \
```
Agreed

amaanq commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
+    assert(tmp != NULL);                                                       \
+    (vec).data = tmp;                                                          \
+    memset((vec).data + (vec).len, 0,                                          \
+           (((_cap) + 1) - (vec).len) * sizeof((vec).data[0]));                \
```
from `man realloc`
> If the new size is larger than the old size, the added memory will not be initialized.

314eter replied:
It's only necessary because you're trying to support null-terminated strings.

amaanq commented on common/scanner.h:
```diff
@@ -1,294 +1,444 @@
+        if (lexer->lookahead != '"') {
+            return false;
+        }
+		advance(lexer);
```
It caused an infinite loop bug, with a hanging opening quote followed by an eof (iirc) never being advanced and it looping back to this section indefinitely

I have to ask then, are comments in line number directives meant to be parsed?

314eter replied:
It looked like a mistake because of the weird indentation. There is indeed a bug here, and adding `advance(lexer)` is probably correct, but it's difficult to review if it's combined with other refactoring.

amaanq commented:
> It's not just a rewrite, you made some other changes too. Some of them look accidental, and some of them don't. It's difficult to review if you don't separate the rewrite from the other changes. Something like this would be ideal:
commit 1: replace stdlib string with custom implementation
commit 2: rewrite C++ to C
commit 3, 4, ...: other fixes and improvements

I apologize for that - I guess I should've split the commits a bit better. Though I can describe everything here.

I fuzzed the scanner with libfuzzer and found several spots for crashes. If you'd like to try it out, you may with this script:

```sh
#!/bin/sh
ROOT_DIR="fuzzer"

LANG=$1
TIME=$2

shift 2

export PATH="/root/.cargo/bin:$PATH"
export CFLAGS="$(pkg-config --cflags --libs tree-sitter) -O0 -g -Wall"

JQ_FILTER='.. | if .type? == "STRING" or (.type? == "ALIAS" and .named? == false) then .value else null end'

build_dict() {
	cat src/grammar.json | jq "$JQ_FILTER" |
		grep -v "\\\\" | grep -v null >$ROOT_DIR/dict
}

build_fuzzer() {
	cat <<END | clang -fsanitize=fuzzer,address $CFLAGS -lstdc++ -g -x c - src/scanner.c src/parser.c $@ -o $ROOT_DIR/fuzzer
#include <stdio.h>
#include <stdlib.h>
#include <tree_sitter/api.h>

TSLanguage *tree_sitter_$LANG();

int LLVMFuzzerTestOneInput(const uint8_t * data, const size_t len) {
  // Create a parser.
  TSParser *parser = ts_parser_new();

  // Set the parser's language.
  ts_parser_set_language(parser, tree_sitter_$LANG());

  // Build a syntax tree based on source code stored in a string.
  TSTree *tree = ts_parser_parse_string(
    parser,
    NULL,
    (const char *)data,
    len
  );
  // Free all of the heap-allocated memory.
  ts_tree_delete(tree);
  ts_parser_delete(parser);
  return 0;
}
END
}

generate_fuzzer() {
	tree-sitter generate
}

makedirs() {
	mkdir -p "$ROOT_DIR"
	mkdir -p "$ROOT_DIR/out"
}

makedirs
# generate_fuzzer

build_dict
build_fuzzer $@
cd "$ROOT_DIR"
./fuzzer -dict=dict -max_total_time=$TIME out/
```

and run ./fuzz.sh ocaml/ocaml_interface 1 60 c (1 being max timeout in seconds for hangs, 60 being fuzz time, c being external scanner lang, c or cpp)

So, when a bug is found, the output is as follow:

![image](https://github.com/tree-sitter/tree-sitter-ocaml/assets/29718261/7af74ea0-9e1c-46e3-bdd2-131515644265)

And a reproducer is written to a file



Next, I also refactored spots where else ifs were unnecessary because the prior if statement had a return or break statement - it is a simple clang-tidy lint to make code more readable, and I also added braces around single-statement blocks, sorry if that added visual bloat to the commit changes

amaanq commented:
I refactored the common scanner into a header and implementation file so that we don't get functions defined twice when compiling both scanners

314eter commented:
> I apologize for that - I guess I should've split the commits a bit better. Though I can describe everything here.

I'd prefer it if you split the commits. You're description is helpful, but it doesn't give a complete overview of which line was changed why. You made some changes to the line number directive scanning based on incorrect assumptions that I noticed by chance, but I'm not confident there aren't any other changes like that.

But more important: I didn't get a reply on this remark yet:

> Can you point to any example/discussion where C++ scanners create problems? This rewrite adds complexity, which should be justified. This doesn't mean I don't want to accept a rewrite, I just want to be sure there is a good reason to do it. If someone comes along in the future with a "use C++ stdlib" PR, why should we reject it?

amaanq commented:
Yes, that's what I just came to as a solution (though as SCANNER_H, I'll do your suggestion instead) - I'll also split the commits shortly

> But more important: I didn't get a reply on this remark yet:
> 
> > Can you point to any example/discussion where C++ scanners create problems? This rewrite adds complexity, which should be justified. This doesn't mean I don't want to accept a rewrite, I just want to be sure there is a good reason to do it. If someone comes along in the future with a "use C++ stdlib" PR, why should we reject it?

C++ in itself doesn't create problems, but there are two problems caused indirectly by using it

One is the list here must be expanded if a grammar wants to have functional wasm bindings:
https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/exports.json#L7

Two is that projects using a pure C toolchain cannot bundle it - like Neovim, which has included python and bash recently due to the scanner rewrites. It really just improves portability and ease of compilation like I mentioned before

Now, are those valid reasons to switch? To me, yes, but to some, it might not be. That being said, I've rewritten over 10 scanners in an effort to make C the default scanner language unless the logic is extremely complex to necessitate needing C++ (though I don't see what would need that, like html's usage of maps was totally unnecessary and could've been replicated with a simple static struct array of arrays where the inner array is the key/value pair in index 0/1 respectively, Max was very appreciative of that)

amaanq commented:
Got it to work with fuzzing & rust bindings just fine, I'll split the commits now

amaanq commented:
Split accordingly now - hopefully it's more readable

314eter commented:
@amaanq I created a PR with an alternative string implementation: https://github.com/tree-sitter/tree-sitter-ocaml/pull/79. The ideas are the same, but it feels a bit cleaner in my opinion.

When this is merged, the C rewrite diff can be very small.

amaanq commented:
I see, that's fine to keep the diff more manageable and I do think inline functions are more readable for this

314eter commented:
I formatted all code with clang-format on master, and reordered the C++ methods to be in the same order as the C functions should be. I rebased this branch on master, and the diff is very clean now.

@amaanq Is there anything else you still want to change, or can we merge it like this?

amaanq commented:
I think a ci action for fuzzing would be nice to add, but that's about it. I can add it in an hour or so or you can copy the file from python or bash and just change the language name, it's in .github/workflows

314eter commented:
I'll merge this then, and add the fuzzer in another PR. Thanks for your work.
